Fix Crash when Instead of View Trigger calls itself

deepakshi-mittal opened the pull request:
### Description

This change will fix crashes when an
Instead of Trigger on View calls itself directly and indirectly
Also fix crash IO view trigger inserted is used with join (BABEL-4514)

Task: BABEL-2170

Engine PR: https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/245

### Issues Resolved

BABEL-2170

### Test Scenarios Covered ###
* **Use case based -**
- IOT Trigger A on view calls Trigger A
- fix crash IO view trigger inserted is used with join (BABEL-4514)

### Check List
- [x ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

deepakshi-mittal committed:
Fix Crash when Instead of View Trigger calls itself

This change will fix crashes when an
Instead of Trigger on View calls itself directly and indirectly

Task: BABEL-2170

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

deepakshi-mittal committed:
Empty-Commit

deepakshi-mittal committed:
Empty-Commit

forestkeeper commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -720,6 +720,28 @@ static bool
+			Trigger *trigger = &trigDesc->triggers[i];
+			Oid current_tgoid = trigger->tgoid;
+			Oid prev_tgoid = InvalidOid;
+			prev_tgoid = lfirst_oid(list_tail(triggerOids));
```
Where is the place add oid into triggerOids ?

deepakshi-mittal replied:
Added in backend code
https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/245/files
src/backend/commands/trigger.c

deepakshi-mittal committed:
Fix for failing test case

Task: BABEL-2170

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

deepakshi-mittal committed:
Added case for indirect recusion

Task: BABEL-2170

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

deepakshi-mittal committed:
Crash fix IO view trigger inserted is used with join

This commit will solve crash script listed in BABEL-4514
When inserted table is used with join inside trigger body,
Instead of Trigger on View was causing server crash

Task: BABEL-2170

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

lejaokri commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -720,6 +720,28 @@ static bool
+			else if (list_member_oid(triggerOids, current_tgoid))
+            {
+                /** Indirect recursive trigger case*/
+                ereport(ERROR,
```
Test this codepath. 

Using the following
```
--setup
create table tab_base ( a int, b int )
go
create view vw_get_base as select * from tab_base
go
create table tab_base2 ( a int, b int )
go
create view vw_get_base2 as select * from tab_base2
go
create trigger tr_vw_get_base_instead_insert
on vw_get_base
INSTEAD OF INSERT
AS
INSERT INTO vw_get_base2 SELECT a, b+1000000 FROM INSERTED
go
create trigger tr_vw_get_base_instead_insert2
on vw_get_base2
INSTEAD OF INSERT
AS
INSERT INTO vw_get_base SELECT a, b+10000 FROM INSERTED
go

```

```
-- error case 

-- Next command will throw error as expected.
-- By the time we get to pltsql_bbfViewHasInsteadofTrigger, triggerOids could have more than one entry.
-- And one of the entries is current_tgoid, so we error out and do error handling
-- But the error handling in ExecCallTriggerFunc will only remove one item from the array.
insert into vw_get_base2 values (2000,1) 
GO

-- This one will run with non-emptry triggerOid because you only removed one from the previous error cleanup.
-- I dont think that's correct.
 insert into vw_get_base values (2000,1) 
GO
```

deepakshi-mittal replied:
I checked this, the triggerOids is being cleaned up properly. The issue is not because of triggerOids variable. Its because the second query which crashes doesn't calls RewriteHandler.

lejaokri replied:
From what I can see, on the first run which is `insert into vw_get_base2 values (2000,1) `, it goes to line 4706 in `exec_stmt_execsql` because `expr->plan == NULL. But on the second run, `expr-plan` is not NULL anymore so it skips the if block

deepakshi-mittal committed:
Added function to fix crash with indirect recursion happening previously

The query plan might be cached and cant rely on erroring out iin
analyzer phase.

Task: BABEL-2170

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

deepakshi-mittal committed:
Revert "Added function to fix crash with indirect recursion happening previously"

This reverts commit ce1eff3f395c5abaa6725b227dc7baff98cf1497.

deepakshi-mittal committed:
Removed indirect recuriosn test case, will fix it seperately

Task: BABEL-2170

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>
